crates/bevy_pbr/src/render/pbr_lighting.wgsl

```diff
@@ -440,10 +440,93 @@ fn perceptualRoughnessToRoughness(perceptualRoughness: f32) -> f32 {
     return clampedPerceptualRoughness * clampedPerceptualRoughness;
 }
 
+// this must align with CubemapLayout in decal/clustered.rs
+const CUBEMAP_TYPE_CROSS_VERTICAL: u32 = 0;
+const CUBEMAP_TYPE_CROSS_HORIZONTAL: u32 = 1;
+const CUBEMAP_TYPE_SEQUENCE_VERTICAL: u32 = 2;
+const CUBEMAP_TYPE_SEQUENCE_HORIZONTAL: u32 = 3;
+
+const X_PLUS: u32 = 0;
+const X_MINUS: u32 = 1;
+const Y_PLUS: u32 = 2;
+const Y_MINUS: u32 = 3;
+const Z_MINUS: u32 = 4;
+const Z_PLUS: u32 = 5;
+
+fn cubemap_uv(direction: vec3<f32>, cubemap_type: u32) -> vec2<f32> {
+    let abs_direction = abs(direction);
+    let max_axis = max(abs_direction.x, max(abs_direction.y, abs_direction.z));
+
+    let face_index = select(
+        select(X_PLUS, X_MINUS, direction.x < 0.0),
+        select(
+            select(Y_PLUS, Y_MINUS, direction.y < 0.0),
+            select(Z_PLUS, Z_MINUS, direction.z < 0.0),
+            max_axis != abs_direction.y
+        ),
+        max_axis != abs_direction.x
+    );
+    
+    var face_uv: vec2<f32>;
+    var corner_uv: vec2<u32> = vec2(0, 0);
+    var face_size: vec2<f32>;
+
+    switch face_index {
+        case X_PLUS:  { face_uv = vec2<f32>(direction.z, -direction.y) / direction.x; }
+        case X_MINUS: { face_uv = vec2<f32>(-direction.z, -direction.y) / -direction.x; }
+        case Y_PLUS:  { face_uv = vec2<f32>(direction.x,  -direction.z) / direction.y; }
+        case Y_MINUS: { face_uv = vec2<f32>(direction.x, direction.z) / -direction.y; }
+        case Z_PLUS:  { face_uv = vec2<f32>(direction.x, direction.y) / direction.z; }
+        case Z_MINUS: { face_uv = vec2<f32>(direction.x, -direction.y) / -direction.z; }
+        default: {}
+    }
+
+    face_uv = face_uv * 0.5 + 0.5;
+
+    switch cubemap_type {
+        case CUBEMAP_TYPE_CROSS_VERTICAL: { 
+            face_size = vec2(1.0/3.0, 1.0/4.0); 
+            switch face_index {
+                case X_PLUS:  { corner_uv = vec2(2, 1); }
+                case X_MINUS: { corner_uv = vec2(0, 1); }
+                case Y_PLUS:  { corner_uv = vec2(1, 0); }
+                case Y_MINUS: { corner_uv = vec2(1, 2); }
+                case Z_PLUS:  { corner_uv = vec2(1, 3); }
+                case Z_MINUS: { corner_uv = vec2(1, 1); }
+                default: {}
+            }
+        }
+        case CUBEMAP_TYPE_CROSS_HORIZONTAL: { 
+            face_size = vec2(1.0/4.0, 1.0/3.0); 
+            switch face_index {
+                case X_PLUS:  { corner_uv = vec2(2, 1); }
+                case X_MINUS: { corner_uv = vec2(0, 1); }
+                case Y_PLUS:  { corner_uv = vec2(1, 0); }
+                case Y_MINUS: { corner_uv = vec2(1, 2); }
+                case Z_PLUS:  { corner_uv = vec2(3, 1); }
+                case Z_MINUS: { corner_uv = vec2(1, 1); }
+                default: {}
```

Comment:
Could you factor this out to reduce shader code duplication? You could do `if cubemap_type == CUBEMAP_TYPE_CROSS_VERTICAL || cubemap_type == CUBEMAP_TYPE_CROSS_HORIZONTAL { switch face_index { ... } }` above the switch. You could also replace the entire switch with an index into a constant array, avoiding all the divergence.

Reply:
Last I knew array indexing had to be by constants, but I’ll double check

Reply:
Oh yes, you're right. But you could do something like `corner_uv = (vec2(0x111102, 0x132011) >> (4 * face_index)) & 0xf`. Up to you, I know that's ugly, but it'd be fast :)

Reply:
that's horrifying. done.